File: src/headless_browser.rs
use headless_chrome::protocol::cdp::types::Event;
use headless_chrome::protocol::cdp::Network;
use headless_chrome::{Browser, LaunchOptionsBuilder};

use regex::Regex;
use reqwest::header;
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tokio::time::sleep;

use crate::GRAFANA_LOCATION;

// url, grafana_dashboard_id, username, password
pub async fn run(grafana_dashboard_id: &str) -> Result<(i32, i32), String> {
    let (cookie_name, cookie_value) = match get_grafana_session_cookie().await {
        Ok((n, v)) => (n.to_string(), v.to_string()),
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    };

    let launch_options = match LaunchOptionsBuilder::default()
        .window_size(Some((1920, 4080))) // Set to 1920x1080 (Full HD)
        .build()
    {
        Ok(v) => v,
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    };

    let browser = match Browser::new(launch_options) {
        Ok(v) => v,
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    };

    let tab = match browser.new_tab() {
        Ok(v) => v,
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    };

    match tab.enable_runtime() {
        Ok(_) => (),
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    }

    match tab.enable_log() {
        Ok(_) => (),
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    }

    let bad_datasource_regex = match Regex::new(r"Datasource \w+ was not found") {
        Ok(v) => v,
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    };

    let nodata_regex = match Regex::new(r"api/ds/query\?ds_type=[^&]+-clickhouse-datasource") {
        Ok(v) => v,
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    };

    let datasource_error_count = Arc::new(Mutex::new(0));
    let datasource_error_count_clone = Arc::clone(&datasource_error_count);

    let nodata_error_count = Arc::new(Mutex::new(0));
    let nodata_error_count_clone = Arc::clone(&nodata_error_count);

    #[allow(clippy::match_single_binding)]
    tab.add_event_listener(Arc::new(move |event: &Event| match event {
        e => {
            let raw_event = format!("{:?}", e.clone());
            if bad_datasource_regex.is_match(&raw_event) {
                if let Ok(mut count) = datasource_error_count_clone.lock() {
                    *count += 1;
                }
            }
            if nodata_regex.is_match(&raw_event) {
                if let Ok(mut count) = nodata_error_count_clone.lock() {
                    *count += 1;
                }
            }
        }
    }))
    .unwrap();

    let cookie = Network::CookieParam {
        name: cookie_name.to_string(),
        value: cookie_value.to_string(),
        url: Some(format!("http://{GRAFANA_LOCATION}/")),
        //domain: Some("host.docker.internal".to_string()),
        domain: None,
        path: Some("/".to_string()),
        secure: Some(false),
        http_only: Some(true),
        same_site: Some(Network::CookieSameSite::Lax),
        expires: None,
        priority: None,
        same_party: Some(false),
        source_scheme: None,
        source_port: None,
        partition_key: None,
    };

    // Set the cookie
    match tab.set_cookies(vec![cookie]) {
        Ok(v) => v,
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    }

    // Navigate to the domain first going back two weeks to test the time-range collar
    let url = format!("http://{GRAFANA_LOCATION}/d/{grafana_dashboard_id}?from=now-14d&to=now");

    let _x = match tab.navigate_to(&url) {
        Ok(v) => v,
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    };

    match tab.wait_until_navigated() {
        Ok(_) => println!("Page navigation completed"),
        Err(e) => return Err(format!("ERROR: {}.{} {e}", file!(), line!())),
    }

    // Wait much longer for all panels to load and make their queries
    println!("Waiting 120 seconds for all panels to load and query data...");
    sleep(Duration::from_secs(120)).await;

    // Reset counters at the start of each iteration
    if let Ok(mut count) = datasource_error_count.lock() {
        *count = 0;
    }
    if let Ok(mut count) = nodata_error_count.lock() {
        *count = 0;
    }

    // Read the counts safely
    let datasource_count = datasource_error_count
        .lock()
        .map(|count| *count)
        .unwrap_or(0);
    let nodata_count = nodata_error_count.lock().map(|count| *count).unwrap_or(0);

    println!(
        "Datasource errors: {}, NoData errors: {}",
        datasource_count, nodata_count
    );

    if datasource_count == 0 && nodata_count == 0 {
        println!("Success! No errors detected.");
    }

    return Ok((datasource_count, nodata_count));
}

async fn get_grafana_session_cookie() -> Result<(String, String), String> {
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none()) // Important: disable redirects
        .build()
        .unwrap();

    let response = client
        .post(format!("http://{GRAFANA_LOCATION}/login"))
        .header(header::CONTENT_TYPE, "application/json")
        .json(&serde_json::json!({
            "user": "admin",
            "password": "admin"
        }))
        .send()
        .await
        .unwrap();

    let cookie = match response
        .headers()
        .get_all(header::SET_COOKIE)
        .iter()
        .next()
        .and_then(|value| value.to_str().ok())
        .ok_or_else(|| format!("ERROR: {}.{} No valid cookie found", file!(), line!()))
    {
        Ok(v) => v,
        Err(e) => {
            return Err(format!(
                "ERROR: {}.{} Invalid cookie format: {e}",
                file!(),
                line!()
            ));
        }
    };

    // Parse cookie name and value
    let cookie_parts: Vec<&str> = cookie.split(';').next().unwrap().split('=').collect();
    if cookie_parts.len() != 2 {
        return Err(format!(
            "ERROR: {}.{} Invalid cookie format",
            file!(),
            line!()
        ));
    }

    Ok((cookie_parts[0].to_string(), cookie_parts[1].to_string()))
}

File: bundle-deployer/src/headless_browser.ts
// Headless browser automation for Grafana dashboard validation

import puppeteer from "puppeteer";
import { GRAFANA_LOCATION } from "./grafana/container.ts";

export async function run(grafanaDashboardId: string): Promise<[number, number]> {
  console.log(`Starting headless browser test for dashboard: ${grafanaDashboardId}`);

  const { cookieName, cookieValue } = await getGrafanaSessionCookie();
  console.log(`Got Grafana session cookie: ${cookieName}`);

  // Try to find system Chrome
  const executablePath = Deno.env.get("PUPPETEER_EXECUTABLE_PATH") ||
    (Deno.build.os === "darwin" ? "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome" :
      Deno.build.os === "linux" ? "/usr/bin/google-chrome" :
        undefined);

  const browser = await puppeteer.launch({
    headless: "new",
    args: ['--no-sandbox', '--disable-setuid-sandbox'],
    executablePath,
  });

  try {
    const page = await browser.newPage();
    await page.setViewport({ width: 1920, height: 4080 });

    let datasourceErrorCount = 0;

    const badDatasourceRegex = /Datasource \w+ was not found/;
    const fourHundredRegex = /400 \w+/;

    // Listen to console messages for datasource errors
    page.on('console', (msg: { text: () => string }) => {
      const text = msg.text();
      console.log("****" + text);
      if (text.includes('400')) {
        datasourceErrorCount++;
        console.log(`ERROR: Datasource not found - ${text}`);
      }
    });

    // Set the session cookie
    await page.setCookie({
      name: cookieName,
      value: cookieValue,
      url: `http://${GRAFANA_LOCATION}/`,
      path: '/',
      httpOnly: true,
      secure: false,
      sameSite: 'Lax',
    });

    const dashboardSlug = await getDashboardSlug(grafanaDashboardId);

    // Navigate to dashboard (2 weeks time range)
    const url = `http://${GRAFANA_LOCATION}/d/${grafanaDashboardId}/${dashboardSlug}`;
    console.log(`Dashboard ID=${grafanaDashboardId}`);
    console.log(`Navigating to: ${url}`);
    await page.goto(url, {
      waitUntil: 'networkidle2',
      timeout: 120000,
    });

    const pageTitle = await page.title();
    const pageUrl = page.url();
    console.log(`Page loaded - Title: "${pageTitle}", URL: ${pageUrl}`);

    console.log("Page navigation completed");

    // Wait for all panels to load and query data
    console.log("Waiting 120 seconds for all panels to load and query data...");
    await new Promise(resolve => setTimeout(resolve, 30000));

    console.log(`Datasource errors: ${datasourceErrorCount}`);

    if (datasourceErrorCount === 0) {
      console.log("Success! No datasource errors detected.");
    }

    // Return 0 for nodata errors since we're not checking those anymore
    return [datasourceErrorCount, 0];

  } finally {
    await browser.close();
  }
}
async function getDashboardSlug(dashboardId: string): Promise<string> {
  const response = await fetch(
    `http://${GRAFANA_LOCATION}/api/dashboards/uid/${dashboardId}`,
    {
      headers: {
        'Authorization': 'Basic ' + btoa('admin:admin'),
      },
    }
  );

  const data = await response.json();
  return data.meta?.slug || 'dashboard';  // Fallback to 'dashboard'
}

async function getGrafanaSessionCookie(): Promise<{ cookieName: string; cookieValue: string }> {
  const response = await fetch(`http://${GRAFANA_LOCATION}/login`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      user: 'admin',
      password: 'admin',
    }),
    redirect: 'manual', // Don't follow redirects
  });

  const setCookieHeader = response.headers.get('set-cookie');

  if (!setCookieHeader) {
    throw new Error("No valid cookie found in response");
  }

  // Parse cookie name and value
  const cookieFirst = setCookieHeader.split(';')[0];
  const cookieParts = cookieFirst.split('=');

  if (cookieParts.length !== 2) {
    throw new Error("Invalid cookie format");
  }

  return {
    cookieName: cookieParts[0],
    cookieValue: cookieParts[1],
  };
}